Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,461 @@
+def Matrix.coxeterSystemCoxeterGroup : CoxeterSystem M M.CoxeterGroup :=
+  CoxeterSystem.ofMulEquiv (MulEquiv.refl _)
+
+/-- A group is a Coxeter group if it is registered in a Coxeter system
+ for some Coxeter matrix `M`. -/
+class IsCoxeterGroup (W : Type u) [Group W] : Prop where
+  nonempty_system : ∃ (B : Type u), ∃ (M : Matrix B B ℕ),
+    M.IsCoxeter ∧ Nonempty (CoxeterSystem M W)
+
+namespace CoxeterSystem
+
+open Matrix
+
+variable {B B' W H : Type*} [Group W] [Group H]
+
+variable {M : Matrix B B ℕ} [M.IsCoxeter]
+
+/-- A Coxeter system for `W` with Coxeter matrix `M` indexed by `B`, is associated to
+a map `B → W` recording the images of the indices. -/
+instance funLike : FunLike (CoxeterSystem M W) B (fun _ => W) where
+  coe cs := fun b => cs.mulEquiv.symm (.of b)
+  coe_injective' := by
+    rintro ⟨cs⟩ ⟨cs'⟩ hcs'
+    have H : (cs.symm : CoxeterGroup M →* W) = (cs'.symm : CoxeterGroup M →* W) := by
+      unfold CoxeterGroup
+      ext i; exact congr_fun hcs' i
+    have : cs.symm = cs'.symm := by ext x; exact FunLike.congr_fun H x
+    rw [ofMulEquiv.injEq, ← MulEquiv.symm_symm cs, ← MulEquiv.symm_symm cs', this]
+
+@[simp]
+theorem mulEquiv_apply_coe (cs : CoxeterSystem M W) (b : B) :
+    cs.mulEquiv (cs b) = .of b :=
+  (MulEquiv.eq_symm_apply cs.mulEquiv).mp rfl
+
+/-- The map sending a Coxeter system to its associated map `B → W`. -/
+theorem ext' {c d : CoxeterSystem M W} (H : ⇑c = ⇑d) : c = d := FunLike.coe_injective H
+
+/-- Extensionality rule for Coxeter systems. -/
+theorem ext {c d : CoxeterSystem M W} (H : ∀ x, c x = d x) : c = d :=
+  ext' <| by ext; apply H
+
+/-- The canonical Coxeter system of the Coxeter group over `X`. -/
+def ofCoxeterGroup (X : Type*) (D : Matrix X X ℕ) : CoxeterSystem D (CoxeterGroup D) :=
+  ofMulEquiv (MulEquiv.refl _)
+
+@[simp]
+theorem ofCoxeterGroup_apply {X : Type*} (D : Matrix X X ℕ) (x : X) :
+    CoxeterSystem.ofCoxeterGroup X D x = CoxeterGroup.of D x :=
+  rfl
+
+@[simp]
+theorem map_relations_eq_reindex_relations (e : B ≃ B') :
+    (MulEquiv.toMonoidHom (FreeGroup.freeGroupCongr e)) '' CoxeterGroup.Relations.toSet M =
+    CoxeterGroup.Relations.toSet (reindex e e M) := by
+  simp [CoxeterGroup.Relations.toSet, CoxeterGroup.Relations.ofMatrix]
+  apply le_antisymm
+  · rw [Set.le_iff_subset]; intro _
+    simp only [Set.mem_image, Set.mem_range, Prod.exists, Function.uncurry_apply_pair,
+      forall_exists_index, and_imp]
+    intro _ hb b _ heq; rw [← heq]
+    use (e hb); use (e b); aesop
+  · rw [Set.le_iff_subset]; intro hb'
+    simp only [Set.mem_range, Prod.exists, Function.uncurry_apply_pair, Set.mem_image,
+      forall_exists_index]
+    intro b1' b2' heq; rw [← heq]
+    use ((FreeGroup.freeGroupCongr e).symm hb')
+    exact ⟨by use (e.symm b1'); use (e.symm b2'); aesop, by aesop⟩
+
+/-- Coxeter groups of isomorphic types are isomorphic. -/
+def equivCoxeterGroup (e : B ≃ B') : CoxeterGroup M ≃* CoxeterGroup (reindex e e M) :=
```

Comment:
```suggestion
@[simps]
def equivCoxeterGroup (e : B ≃ B') : CoxeterGroup M ≃* CoxeterGroup (reindex e e M) :=
```
Hopefully this will autogenerate the simp-lemmas below.

Reply:
It wants @[simps!] so will use that.